Capstone-Projects-2023-Spring/project-virtual-pet
# Sb 24 second onesignal notifications

Yes, finally!

### The tl;dr:
I tested this on localhost and then again by deploying this branch. It works on Safari consistently, but see below for why Chrome is not vibing with OneSignal. I tested this with an iPhone, too, so it works on mobile! 

Here's a screenshots of it working on Chrome from my Mac:
<img width="1207" alt="Screen Shot 2023-04-26 at 1 24 37 AM" src="https://user-images.githubusercontent.com/73796086/234477381-e0c1a563-0847-4f26-976d-adf6e00c5608.png">

### **TO TEST IT ON LOCALHOST:**
1. To test locally, you need to actually change a setting in OneSignal. Please ping me if you'd like to do this because I need to go into my account to do this. 
<img width="1122" alt="Screen Shot 2023-04-26 at 1 23 16 AM" src="https://user-images.githubusercontent.com/73796086/234477192-a29a5f47-73ef-4665-a962-d58f0ff5a251.png">
2. Change this line in Main.js `allowLocalhostAsSecureOrigin: false, // change to true for testing on localhost, false for https site` you need to change it from `false` to `true`.
3. Make sure you've cleared your cache and that your notifications are set to default. 
4. Make sure you aren't using Private Browsing Mode / Incognito mode on Chrome or Chrome's full screen mode as this can hide and disable notifications. Make sure you do NOT have focus mode on or do not disturb.

### **TO TEST ON THE SITE:**
1. Once deployed to the site, you should be able to test notifications. Make sure you've cleared your cache and that your notifications are set to default. 

### **A GOOD WAY TO TEST ONCE YOU'RE SET UP:**
1. Navigate to the site. Login. There should be a bell icon in the bottom right corner.
2. Click it. The native browser notification subscription prompt should come up. Click `Accept`.
3. A welcome notifications should pop up.
6. Open developer tools on your browser and in the Console, type `OneSignal.log.setLevel('trace');` to enable web SDK logging. You should see `undefined` as the result.
7. Close the tab and open a new tab to the same page (refreshing the page is not enough to trigger some SDK events). 
8. Upon coming back to the page and reopening the Console, You should see a lot of output which will narrate initializing OneSignal, installing service worker, enabling subscriptions, etc. You shouldn't see any errors.
9. You can disable this additional logging by entering `OneSignal.log.setLevel('warn');` when you're done.
10. Now enter 
```
function isPushNotificationsEnabledVerbose() {
    console.log('isPushNotificationsEnabledVerbose()');
    Promise.all([
            OneSignal.isPushNotificationsEnabled(),
            OneSignal.getUserId(),
            OneSignal.getRegistrationId(),
            OneSignal.getNotificationPermission(),
            OneSignal.isOptedOut(),
            OneSignal.context.serviceWorkerManager.getActiveState()
        ])
        .then(([isSubscribed, userId, registrationId, notificationPermission, optedOut, serviceWorkerActive]) => {
            console.log('Is Completely Subscribed:', isSubscribed);
            console.log('');
            console.log('What is our OneSignal user ID?', userId);
            console.log('What is our push subscription token?', registrationId);
            console.log('What is the notification permission status?', notificationPermission);
            console.log('Are you manually opted out?', optedOut);
            console.log("Is a service worker registered and active? (should be false on Safari, otherwise should be 'Worker A (Main)')?", serviceWorkerActive);
            console.log('What is the current URL of this page?', location.href);
            console.log("What environment does OneSignal think it's in?", OneSignal.sdkEnvironment.getWindowEnv());
        })
        .catch(e => {
            console.error("Issue determining whether push is enabled:", e);
        });
}
isPushNotificationsEnabledVerbose();
```
11. Send yourself a test notification by entering `OneSignal.sendSelfNotification()` in the console. You should see something like to `Promise {[[PromiseStatus]]: "pending", [[PromiseValue]]: undefined}`, and then a web push notification should pop up. 

### Resources (for troubleshooting, etc):
I used [this](https://documentation.onesignal.com/docs/onesignal-service-worker-faq#service-worker-migration-guide) and [this]() to help me merge our existing service worker with the OneSignal SDK one. Then I followed some advice from [this blogpost](https://onesignal.com/blog/how-to-integrate-push-notifications-in-react/) and [this webinar](https://onesignal.com/videos/integrate-push-notifications-into-your-reactjs-apps) to incorporate notification capabilities.  The reply to [this StackOverflow post](https://stackoverflow.com/questions/68634626/converting-existing-react-app-to-pwa-that-is-using-onesignal-and-react-app-rewir) was particularly helpful and even directed me to a Discord forum where I asked questions and found info from other developers. [This page](https://documentation.onesignal.com/docs/troubleshooting-web-push#debugging-using-browser-developer-tools) was AWESOME for helping to troubleshoot, and you can even copy/paste some stuff from here for testing purposes.

### A note about clearing cache/site data:
I have found that there are so many issues with caching that can mess you up if you're testing notifications over and over, subscribing/unsubscribing/resubscribing, etc. I learned this the hard way... You NEED to clear your cache and site data if you are testing these aspects on your browser, whether locally or on the deployed site. [Heres good instructions for it.](https://documentation.onesignal.com/docs/clearing-cache-and-resetting-push-permissions#reset-safari-on-macos)

### A note about Chrome on Mac:
Due to a [bug with OneSignal](https://github.com/OneSignal/OneSignal-Website-SDK/issues/857), it seems like there are some issues with newer versions of Chrome on Mac (where multiple notifications from the same app are grouped by the same ID, so only the first is shown, then subsequent notifications do not come up as alerts). However, if you click the clock in the upper right corner of your Mac screen to open the Notification Center, the notifications should be there, and the browser recognizes them as received, and OneSignal recognizes them as delivered. If you clear your notifications, the next message *will* alert you. I'm not prioritizing fixing this, and have contacted the OneSignal team about this issue.
